Remove nvim-treesitter dependency

TheLeoP opened the pull request:
I started working on this without realizing that #330 existed :p

This PR:
- Removes the nvim-treesitter dependency (and hence, it works perfectly with its `main` branch).
- Only memoizes the computed matches using the buffer, language and id of the node (tied to it's state) as hash. Using buftick has proven to cause issues like in https://github.com/nvim-treesitter/nvim-treesitter/issues/4902 and other function do not have access to the id of the root of the tree, which is the hashable value that tells us if the tree of a buffer has changed.
- Adds a bunch of lua annotations to ease contributing to the Lua side of the codebase.
- Cleans the lua `setup` wrapper (that simply defines `g:` vimscript variables).
- Uses a single interface for configuration (`g:` vimscript variables).
- Removes the dependency on the Neovim highlighter (outside of the `syntax.lua` module). Now this plugin parses the buffer if necesary (and could even use the `g:matchup_delim_stopline` if we could know which window is being used to get the cursor location. An option would be to always use the current window, but I don't know enough of the codebase to say if that would work).
- format all queries using https://github.com/ribru17/ts_query_ls
- remove the dependency on the old `make-range` directive. Instead, it relies on the fact that treesitter matches can capture multiple nodes under the same name. So, a range can be obtained from the first and last nodes with the same name on a given match
- adds a small query to use treesitter to `%` between the start and end of Lua tables because I was getting mad at an edge case that the regex based engine couldn't catch on a file in my personal config (https://github.com/TheLeoP/nvim-config/blob/1ba78ab0a268fd5f8e3dc371a82e1af44c829fee/lua/personal/emmet.lua#L27-L51)


Just `memoize` was copied from `nvim-treesitter`, the whole `query.lua` dependency was removed in favor of a smaller (and non-nvim-treesitter dependant) implementation of the functions used by this plugin.

This PR does not
- modify tests
- modify Github actions
- add a Dockerfile for tests
- add a Makefile for creating a docker image for tests

This PR lacks documentation for the changes introduced in it, but I wanted to see what everyone thinks of it before commit further to it.

Additionally, since treesitter parsing can now be async, I could make use of it in follow up PR if deemed necessary.

andymass commented:
Thank you for contributing 🙏!  Looks like are some merits to both this and #330, I will take bits of both.  Since they are extensive I will likely merge them in bits and pieces where possible.

Both PRs have mentioned performance issues (which likely do exist in head already) and which are not yet solved by either.

https://github.com/andymass/vim-matchup/pull/330#issuecomment-2909856135

> Additionally, since treesitter parsing can now be async, I could make use of it in follow up PR if deemed necessary.

I am curious which approach do we think is more viable?  Can you give any more detail on what changes would be required for async parsing (no need for a full PR for that yet)?

andymass commented:
@Slotos any concerns to go ahead and merge this PR?  Based on your comments regarding "cleanup around old attach/detach routines" it seems this will be a bit cleaner to build on top of.  I would like to take your CI/CD updates though.  Just want to make sure any performance improvements you are planning could be built on top of this CR and not make it difficult for you.

@TheLeoP Any guidance for users to migrate?  It seems they just need to replace a call from nvim-treesitter.setup to matchup.setup, I will have to add some readme section on this migration.  Doing nothing it seems the user's config would break right?

TheLeoP commented:
> Any guidance for users to migrate? 

Users simply need to add

```lua
vim.g.matchup_treesitter_enabled = true
```

or

```vim
g:matchup_treesitter_enabled = v:true
```

to their configs. They could also use the lua interface if they want to

```lua
require("match-up").setup {
  treesitter = {
    enabled = true,
  },
}
```

It's worth noting that I'm using either Lua's `true` and `false` or Neovims `v:true` and `v:fase` instead of `1` and `0` like on other options.

> Doing nothing it seems the user's config would break right?

It shouldn't break, unless I'm missing something. It'll stop using treesitter by default and rely on the regex engine unless they enable treesitter as mentioned above. I don't know how `nvim-treesitter` on their `master` branch handles the configuration of non-existent modules, thought. So I'm not sure if that may throw errors for users configurations.

andymass commented:
> I'm a strong opponent of require.setup.. Editor plugins should just work

I agree- the treesitter capability should be enabled by default now and controlled just by g: variables (like in @TheLeoP's PR) as this plugin has for every other functionality.  However, I think we should still offer an optional setup function which simply updates the corresponding g: variables because many neovim users expect this convenience.

This would mean setting `vim.g.matchup_treesitter_enabled = true` by default.

> It also works correctly with injections, whereas parse(nil)

Thanks, got it, we need @Slotos's version of this part.

TheLeoP commented:
@peter-bread  that's not because of this PR (as far as I can tell). That happens simply because there is no treesitter capture for a function call and the regex engine fails to correctly find the closing parenthesis.

If you add the following capture to `after/queries/lua/matchup.scm`, it works as expected

```scheme
(function_call
  (arguments
    "(" @open.call
    ")" @close.call)) @scope.call
```

TheLeoP commented:
Thanks for the example @peter-bread , I now understand the issue properly.

The issue was an oversight on my side when modifying the vimscript side of the codebase. It should be working in the same was as `master` or #330 in this particular case.

andymass commented on autoload/matchup.vim:
```diff
@@ -77,7 +77,7 @@ function! s:init_options()
   call s:init_option('matchup_matchpref', {})
 
-  call s:init_option('matchup_treesitter_enabled', v:false)
+  call s:init_option('matchup_treesitter_enabled', v:true)
```
could you put this behind a if has('neovim')?

TheLeoP commented:
> Is this a good enough fix?

In my opinion, it looks good enough. I also don't know why the builtin `gcc` is creating unloaded buffers, but given that treesitter parsers can't be created for unloaded buffers, this check seems appropriate. 

> Should vim-matchup be more selective when checking buffers (e.g. should it be checking buffers that are 'nofile' or hidden or unnamed)?

I lack context of the bigger `vim-matchup` codebase, specially on the vimscript side. But that seems reasonable.

It's strange that the `Filetype` autocmd is being triggered in an unloaded buffer, though. Maybe worth further investigation and/or opening an issue in the Neovim repo itself. Specially because this is being triggered by the builtin commenting functionality.

andymass commented:
Thanks for your extensive and continued work @TheLeoP!  Do you think you are you able to handle the key issues @Slotos raised (potentially using code from @Slotos ’s branch)?  (Note I think the only blocking one at this point is regarding parsing and injections)

TheLeoP commented:
> Thanks for your extensive and continued work @TheLeoP! Do you think you are you able to handle the key issues @Slotos raised (potentially using code from @Slotos ’s branch)? (Note I think the only blocking one at this point is regarding parsing and injections)

Of course. I'll be working on doing it

TheLeoP commented:
The parsing issue should be fixed now. I also added documentation for the new `g:matchup_treesitter` variables and modified the README a bit. I noticed that my autoformatter changed more of the README than I intended to, so let me know if that's an issue in order to fix it @andymass .

I chose `400` as the default value for `g:matchup_treesitter_stopline` because it may have a high impact in performance and it's already the default value of `g:matchup_matchparen_stopline`.

Slotos commented on lua/treesitter-matchup/internal.lua:
```diff
@@ -1,158 +1,266 @@
       local lang = lang_tree:lang()
-      local group_results = queries.collect_group_results(
-        bufnr, 'matchup', tree:root(), lang) or {}
+      local group_results = get_memoized_matches(bufnr, tree:root(), lang)
```
Here be dragons.

Memoization is done on language tree status, but results of memoized function are dependent on cursor position, or rather tree range being queried. It depends on queries too, but that's solvable by wiping memoized cache on queries update.

I just tries this branch on a 2000+ JS file, and treesitter backend is effectively disabled past first 200 lines.

I'd recommend to forgo optimization attempts here, this is not a problem you've introduced.

TheLeoP replied:
A quick profile in a 900 lines typescript file with a 400-ish lines function (jumping between two ends of the function once than range has been parsed. So, not the first time) shows that everything is working as expected functionality wise. However, it seems like `parser:parse({start_row, end_row})` always takes longer that `parser:parse(true)` for an unchanged buffer. Using `g:__ts_debug` to see how much time is taking the treesitter parsing of the buffer, shows that is taking 0 ms. My guess is that there is some overhead in the Lua side when checking for valid regions one by one, but I haven't been able to fully confirm this. When using a debugger, it seems to always find the tree as valid and return early, so there may be some race condition happening.

The slice flamegraph for `parser:parse({start_row, end_row})` seems like the following (I made multiple jumps. 2s it's not the duration of a single jump)

![image](https://github.com/user-attachments/assets/e33f806d-3542-406b-9540-f7de5e48f48d)

The slice flamegraph for `parser:parse(true)` seems like the following (same)

![image](https://github.com/user-attachments/assets/d378fc57-74da-40af-a499-b7e3173f2769)

@Slotos your comment about the treesitter engine not working after the first 200 lines and this quick example makes me think that what you may be seeing not because of memoization, but instead because you are trying to find matches between not yet parsed regions of the buffer.

andymass commented:
>  Is it ok if I add a bunch of these kind of captures @andymass ?

certainly, no concern here

TheLeoP commented:
I think this PR is ready to be merged/further comments @andymass

Slotos commented:
A bit of self-promotion, docker test runner commit from my obsolete PR can help a lot with local development.

I can’t write code without tests, it was an absolute must-have for me.

TheLeoP commented:
I've rebased my branch on top of @Slotos' docker commit and ran `make docker_test_new` and `make docker_test_old`. In both cases 10 out of 10 test cases succeeded on Neovim 0.11.2 (the latest stable version). The treesitter interface on Neovim had some changes on `0.11` , so I guess that may be what's causing the Github actions test to fail on Neovim 0.10.1

andymass commented:
@TheLeoP  are there likely to be users of neovim 0.10.1 or does everyone use 0.11?  Note that head also seems to have failed.

TheLeoP commented:
Tests are passing now @andymass :D

@slotos I copied most of your changes regarding regarding CI, Docker and some tests changes. I included you as a co-author of the commits (I would have made you the sole author if I could). So, once again, thanks for your help.

TheLeoP commented:
> are there likely to be users of neovim 0.10.1 or does everyone use 0.11?

Not everyone will ever be in the same version. I guess that most non-novices will be using either `0.11.2` or nightly. Latest available Neovim version on stable distros (I checked Ubuntu and Debian) seems to be `0.10`. Using the Neovim appimage (in order to access `0.11.2`) is always an option, though.

Ultimately, the decisions is yours, @andymass . For what is worth, the plugin mentions in both the README and matchup.txt that only the latest stable version of Neovim is supported for the treesitter integration.

andymass commented:
Fantastic, merging!  Thanks again all for the extensive work.

> the plugin mentions in both the README and matchup.txt that only the latest stable version of Neovim is supported for the treesitter integration.

That's true, though I am concerned someone will update the plugin who was not using treesitter, which is now default on, and may get errors.  We'll see if anyone has such issues.
